pkg/cache/cluster_info.go

```diff
@@ -356,24 +344,32 @@ func (m *ClusterInfo) processNewSchedulableNodes(request *si.UpdateRequest) {
         log.Logger().Info("successfully added node",
             zap.String("nodeId", node.NodeId),
             zap.String("partition", nodeInfo.Partition))
+        // create the equivalent scheduling node
+        m.EventHandlers.SchedulerEventHandler.HandleEvent(
+            &schedulerevent.SchedulerNodeEvent{
+                AddedNode: nodeInfo,
```

Comment:
That means that we do not have a new node until we have processed all nodes and then stop scheduling for the time it takes to add all nodes to the cluster. This will not show up during the unit tests but when you add or remove a large number of nodes during normal operations that might lead to longer lock times and no scheduling actions during that time.

The way we do it now the scheduling node *should* be ready when we are done. The time that the cache and the scheduler are out of sync is minimal.
I was even looking at pulling in the existing allocations inside the loop. Cache and scheduler should be independent and could even run from different go routines but they must be in sync as soon as possible.

Reply:
This is not a very strong point to me, because adding a node to cache is not a time-consuming work, we are not expecting any meaningful delay here. Instead, if we can pack them up for event dispatching, it is more efficient (scheduler channel is quite busy handling allocations already) and also the coding style is more consistent.

Reply:
A channel should be able to move millions of objects per second. It is not the channel process or the number of objects that causes the channel to fill up it is the speed at which we pull them off and can process each of them. Doing one node at a time makes processing really fast and keeps lock time as low as possible. Having large numbers of nodes, like the 5,000 we do in the test is exceptional. A new rack of nodes is the largest number (50 nodes or so) I would expect in normal operating circumstances.

The other point to make to keep it at single nodes is: if we are really scheduling we do not want to hold the lock to process all the nodes in one batch. That could cause a large pause in scheduling. This gets a far more even spread and other events can be processed in between.

I would even argue that packing up any type of update and pushing batched updates is the wrong way as it can lead to longer pauses. You want to keep it lightweight, with short lock times using highly granular locks not heavy with long lock time and coarse locking.

Reply:
Make sense to me, thanks!